chain/network/src/concurrency/scope/mod.rs

```diff
@@ -57,254 +57,245 @@ use crate::concurrency::signal;
-/// the concurrency model (you can still implement a workaround by using a channel,
-/// if you really want to, but that might mean that Scope is not what you want
-/// in the first place).
-/// In particular Service::spawn() doesn't return a JoinHandle,
-/// because it can be called outside of the scope that it belongs to.
-pub struct JoinHandle<'env, T>(
-    tokio::task::JoinHandle<Result<T, ErrTaskCanceled>>,
-    std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
-);
 
-#[derive(thiserror::Error, Debug)]
-#[error("awaited task has been canceled")]
-pub struct ErrTaskCanceled;
-
-impl<'env, T> JoinHandle<'env, T> {
-    /// Awaits for a task to be completed and returns the result.
-    /// Returns Err(ErrCanceled) if the awaiting (current task) has been canceled.
-    /// Returns Ok(Err(ErrTaskCanceled) if the awaited (joined task) has been canceled.
-    pub async fn join(self) -> ctx::OrCanceled<Result<T, ErrTaskCanceled>> {
-        ctx::wait(async { self.0.await.unwrap() }).await
+    fn register(&self, err: E) {
+        let mut m = self.0.err.lock().unwrap();
+        if m.is_some() {
+            return;
+        }
+        *m = Some(err);
+        self.0.ctx.cancel();
     }
-}
 
-impl<E: 'static + Send> Inner<E> {
-    /// Spawns a task in the scope, which owns a reference of to the scope, so that scope doesn't
-    /// terminate before all tasks are completed.
+    /// Spawns a task in the scope, which owns a reference of to the scope,
+    /// so that scope doesn't terminate before all tasks are completed.
+    ///
+    /// The reference to the scope can be either `Arc<TerminateGuard>` or `Arc<CancelGuard>`,
+    /// so that the scope may get terminated/canceled when the guard is dropped.
     ///
-    /// The reference to the scope can be an arbitrary
-    /// type, so that a custom drop() behavior can be added. For example, see `StrongService` scope reference,
-    /// which cancels the scope when dropped.
+    /// Returns a handle to the task, which awaits either for the task to return succesfully,
+    /// or for the WHOLE scope to terminate.
     fn spawn<M: 'static + Send + Sync + Borrow<Self>, T: 'static + Send>(
         m: Arc<M>,
         f: impl 'static + Send + Future<Output = Result<T, E>>,
-    ) -> tokio::task::JoinHandle<Result<T, ErrTaskCanceled>> {
+    ) -> tokio::task::JoinHandle<Result<T, Arc<Inner<E>>>> {
         tokio::spawn(must_complete(async move {
-            match (ctx::CtxFuture { ctx: m.as_ref().borrow().ctx.clone(), inner: f }).await {
+            match (ctx::CtxFuture { ctx: m.as_ref().borrow().0.ctx.clone(), inner: f }).await {
                 Ok(v) => Ok(v),
                 Err(err) => {
-                    m.as_ref().borrow().output.send(err);
-                    Err(ErrTaskCanceled)
+                    let guard = m.as_ref().borrow();
+                    guard.register(err);
+                    let inner = guard.0.clone();
```

Comment:
I used just "Inner", because it already lives inside scope module (so more like scope::Inner) but I'm fine with ScopeInner as well (Impl is not really idomatic for rust afaik).

When it comes to tuples: I use tuples with unnamed fields, because all of the fields of the tuple are usually pointers to the same thing, just with a different kind of indirection. This way I combine properties of different wrappers and I believe that making the fields unnamed, kind of suggests that the subscript doesn't really matter (i.e. we refer to the same thing, just via different indirection).